src/scalar_mul/pippenger.rs

```diff
@@ -0,0 +1,211 @@
+    /// 6. Add to the return value.
+    /// 7. Repeat the loop.
+    ///
+    /// Approximate cost w/o wNAF optimizations (A = addition, D = doubling):
+    ///
+    /// ```ascii
+    /// cost = (n*A + 2*(2^w/2)*A + w*D + A)*256/w
+    ///          |          |       |     |   |
+    ///          |          |       |     |   looping over 256/w windows
+    ///          |          |       |     adding to the result
+    ///    sorting points   |       shifting the sum by w bits (to the next window, starting from last window)
+    ///    one by one       |
+    ///    into buckets     adding/subtracting all buckets
+    ///                     multiplied by their indexes
+    ///                     using a sum of intermediate sums
+    /// ```
+    ///
+    /// For large `n`, dominant factor is (n*256/w) additions.
+    /// However, if `w` is too big and `n` is not too big, then `(2^w/2)*A` could dominate.
+    /// Therefore, the optimal choice of `w` grows slowly as `n` grows.
+    ///
+    fn vartime_multiscalar_mul<I, J>(scalars: I, points: J) -> EdwardsPoint
+    where
+        I: IntoIterator,
+        I::Item: Borrow<Scalar>,
+        J: IntoIterator,
+        J::Item: Borrow<EdwardsPoint>,
+    {
+        let mut scalars = scalars.into_iter();
+        let size = scalars.by_ref().size_hint().0;
+        let w = if size < 3 {
+            1
+        } else if size < 13 {
+            2
+        } else if size < 27 {
+            3
+        } else if size < 66 {
+            4
+        } else if size < 190 {
+            5
+        } else if size < 500 {
+            6
+        } else if size < 800 {
+            7
+        } else if size < 3_000 {
+            8
+        } else if size < 6_000 {
+            9
+        } else if size < 13_000 {
+            10
+        } else if size < 23_000 {
+            11
+        } else if size < 45_000 {
+            12
+        } else if size < 160_000 {
+            13
+        } else if size < 260_000 {
+            14
+        } else {
+            15
+        };
+
+        let window_size = 1 << w;
+        let windows_count = (256 + w - 1) / w; // == ceil(256/w)
+        let buckets_count = window_size / 2; // digits are signed+centered hence 2^w/2, excluding 0-th bucket
+
+        // Step 1. Prepare scalar representation using signed digits radix 2^w.
+        // Each digit will be in [-2^w/2, 2^w/2].
+        let all_digits: Vec<_> = scalars
+            .into_iter()
+            .map(|c| c.borrow().to_arbitrary_radix(w))
+            .collect();
+        let ps: Vec<ProjectiveNielsPoint> = points
+            .into_iter()
+            .map(|p| p.borrow().to_projective_niels())
+            .collect();
+
+        // Iterate the digits from last to first so we can shift the result by w bits
+        // at the end of each iteration (e.g. w doublings).
+        let mut result = EdwardsPoint::identity();
+        for i in (0..windows_count).rev() {
```

Comment:
Is `windows_count` supposed to be the number of digits in each scalar, so that `(0..windows_count).rev()` iterates over the digits in reverse order?

Reply:
Yeah. Window <=> digit.